Sources/PointFree/Auth.swift

```diff
@@ -5,14 +5,47 @@ import HttpPipeline
 import HttpPipelineHtmlSupport
 import Optics
 import Prelude
+import UrlFormEncoding
 
 let secretHomeResponse: (Conn<StatusLineOpen, Prelude.Unit>) -> IO<Conn<ResponseEnded, Data>> =
   writeStatus(.ok)
     >-> readGitHubSessionCookieMiddleware
     >-> respond(secretHomeView)
 
 let githubCallbackResponse =
-  authTokenMiddleware
+  extractGitHubAuthCode
+    <| authTokenMiddleware
+
+/// Middleware transformer to convert the optional GitHub code to a non-optional. In the `nil` case we show
+/// a 400 Bad Request page.
+private func extractGitHubAuthCode(
+  _ middleware: @escaping Middleware<StatusLineOpen, ResponseEnded, (code: String, redirect: String?), Data>
+  )
+  -> Middleware<StatusLineOpen, ResponseEnded, (code: String?, redirect: String?), Data> {
+    
+    return { conn in
+      conn.data.code
+        .map { (code: $0, redirect: conn.data.redirect) }
+        .map { conn.map(const($0)) }
+        .map(middleware)
+        ?? (conn |> const(unit) >¢< missingGitHubAuthCodeMiddleware)
+    }
+}
+
+// TODO: Move to HttpPipeline
+private func >¢< <A, B, C, I, J>(
+  lhs: @escaping (A) -> C,
+  rhs: @escaping Middleware<I, J, C, B>
+  )
+  -> Middleware<I, J, A, B> {
+
+    return map(lhs) >>> rhs
+}
+
+/// Middleware to run when the GitHub auth code is missing.
+private let missingGitHubAuthCodeMiddleware: Middleware<StatusLineOpen, ResponseEnded, Prelude.Unit, Data> =
+  writeStatus(.badRequest)
+    >-> respond(text: "GitHub code wasn't found :(")
```

Comment:
can make this response better later. prob some kind of error specific view or somethin